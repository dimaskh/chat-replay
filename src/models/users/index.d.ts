// Props
export declare const PROP_ID: 'id'
export declare const PROP_USER_NAME: 'user_name'
export declare const PROP_USERNAME: 'username'
export declare const PROP_DISPLAY_NAME: 'display_name'

// Model
export declare const MODEL: 'User'

// Entity
export type Message = {
  [PROP_ID]: number
  [PROP_USER_NAME]?: string
  [PROP_USERNAME]?: string
  [PROP_DISPLAY_NAME]: string
}

// Getters
export declare const getId: GetProperty<typeof PROP_ID>
export declare const getUserName: GetProperty<typeof PROP_USER_NAME>
export declare const getDisplayName: GetProperty<typeof PROP_DISPLAY_NAME>
