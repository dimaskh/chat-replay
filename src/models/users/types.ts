// Constants
import { PROP_DISPLAY_NAME, PROP_ID, PROP_USERNAME, PROP_USER_NAME } from './constants'

// Entity
export type User = {
  [PROP_ID]: number
  [PROP_USER_NAME]?: string
  [PROP_USERNAME]?: string
  [PROP_DISPLAY_NAME]: string
}
