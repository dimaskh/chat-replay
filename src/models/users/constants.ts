// Props
export const PROP_ID = 'id'
export const PROP_USER_NAME = 'user_name'
export const PROP_USERNAME = 'username'
export const PROP_DISPLAY_NAME = 'display_name'

// Model
export const MODEL = 'User'
