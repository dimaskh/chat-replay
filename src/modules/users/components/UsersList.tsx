// Libs
import React from 'react'

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'column' as any,
    width: 200,
    height: '100%',
    border: '1px solid grey',
  },
}

export default class UsersList extends React.PureComponent<any> {
  public render() {
    return <div style={styles.container}>Content</div>
  }
}
