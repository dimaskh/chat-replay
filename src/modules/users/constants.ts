// Prefix
export const PREFIX = 'modules/users'

// Actions
export const RECEIVE_USERS = `${PREFIX}/RECEIVE_USERS`
export const REMOVE_USER = `${PREFIX}/REMOVE_USER`
export const UPDATE_USER = `${PREFIX}/UPDATE_USER`
