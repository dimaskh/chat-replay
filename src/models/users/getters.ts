// Composer
import composeGetter from 'utils/composeGetter'

// Constants
import * as props from './constants'

export const getId: Function = composeGetter(props.PROP_ID)
export const getUserName: Function = composeGetter(props.PROP_USER_NAME)
export const getDisplayName: Function = composeGetter(props.PROP_DISPLAY_NAME)
