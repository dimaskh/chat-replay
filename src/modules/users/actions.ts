// Constants
import { RECEIVE_USERS } from './constants'

// Types
import { User } from 'models/users/types'

export const receiveUsers = (users: User[] = []) => ({
  type: RECEIVE_USERS,
  payload: users,
})
