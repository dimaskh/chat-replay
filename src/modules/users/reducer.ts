// Constants
import { PROP_ID } from 'models/users/constants'
import { RECEIVE_USERS, REMOVE_USER, UPDATE_USER } from './constants'

// Utils
import createReducer from 'utils/createReducer'

// Types
import { User } from 'models/users/types'
import { AnyAction } from 'redux'

export type UsersState = {
  data: User[]
}

export const initialState: UsersState = {
  data: [],
}

export const receiveUsers = (state: UsersState, action: AnyAction): UsersState => {
  const { payload: users } = action

  return {
    ...state,
    data: [...state.data, ...users],
  }
}

export const updateUser = (state: UsersState, action: AnyAction): UsersState => {
  const { payload: updatedUser } = action

  const newData = state.data.map((user: User) => {
    if (user[PROP_ID] === updatedUser[PROP_ID]) {
      return updatedUser
    }

    return user
  })

  return {
    ...state,
    data: newData,
  }
}

export const removeUser = (state: UsersState, action: AnyAction): UsersState => {
  const { payload: id } = action

  const newData = state.data.filter((user: User) => user[PROP_ID] !== id)

  return {
    ...state,
    data: newData,
  }
}

export default createReducer(initialState, {
  [RECEIVE_USERS]: receiveUsers,
  [REMOVE_USER]: removeUser,
  [UPDATE_USER]: updateUser,
})
