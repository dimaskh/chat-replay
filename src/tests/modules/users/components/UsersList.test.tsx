// Libs
import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

// Tested component
import UsersList from 'modules/users/components/UsersList'

// Mocks
import { user } from 'tests/mocks/users'

describe('users/UsersList component', () => {
  it('should render correctly', () => {
    const component = shallow(<UsersList users={[user]} />)

    expect(shallowToJson(component)).toMatchSnapshot()
  })
})
