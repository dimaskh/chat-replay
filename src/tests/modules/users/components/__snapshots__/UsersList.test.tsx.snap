// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users/UsersList component should render correctly 1`] = `
<styled.div>
  <Header
    medium={true}
  >
    Users
  </Header>
  <UserItem
    key="1"
    user={
      Object {
        "display_name": "Taco Spolsky",
        "id": 1,
        "user_name": "taco",
      }
    }
  />
</styled.div>
`;
